File: classifier/data/raw/ai/samples/project_euler_zig/problem_038.zig
const std = @import("std");

// Project Euler 38: largest 1-to-9 pandigital concatenated product.
pub fn main() !void {
    var largest: u64 = 0;
    var base: u64 = 1;
    while (base < 10_000) : (base += 1) {
        var concatenated: u64 = 0;
        var multiplier: u64 = 1;
        while (concatenated <= 987_654_321) : (multiplier += 1) {
            const part = base * multiplier;
            concatenated = concatenated * powerOfTen(part) + part;
            if (concatenated > 987_654_321) break;
            if (isPandigital(concatenated) and concatenated > largest) largest = concatenated;
        }
    }
    try printAnswer(largest);
}

fn powerOfTen(number: u64) u64 {
    var power: u64 = 10;
    var n = number;
    while (n >= 10) : (n /= 10) power *= 10;
    return power;
}

fn isPandigital(number: u64) bool {
    var n = number;
    var mask: u16 = 0;
    var digits: u8 = 0;
    while (n > 0) : (n /= 10) {
        const digit: u4 = @intCast(n % 10);
        if (digit == 0 or (mask & (@as(u16, 1) << digit)) != 0) return false;
        mask |= @as(u16, 1) << digit;
        digits += 1;
    }
    return digits == 9 and mask == 0b11_1111_1110;
}

fn printAnswer(value: anytype) !void {
    var buffer: [64]u8 = undefined;
    const text = try std.fmt.bufPrint(&buffer, "{d}\n", .{value});
    try std.Io.File.stdout().writeStreamingAll(std.Options.debug_io, text);
}

File: classifier/data/raw/ai/samples/project_euler_java/Problem038.java
/** Project Euler 38: largest 1-to-9 pandigital concatenated product. */
public class Problem038 {
    public static void main(String[] args) {
        int largest = 0;
        for (int n = 1; n < 10_000; n++) {
            StringBuilder concatenated = new StringBuilder();
            for (int multiplier = 1; concatenated.length() < 9; multiplier++) {
                concatenated.append(n * multiplier);
            }
            if (concatenated.length() == 9 && isPandigital(concatenated)) {
                largest = Math.max(largest, Integer.parseInt(concatenated.toString()));
            }
        }
        System.out.println(largest);
    }

    private static boolean isPandigital(CharSequence value) {
        int mask = 0;
        for (int i = 0; i < value.length(); i++) {
            int digit = value.charAt(i) - '0';
            if (digit == 0 || (mask & (1 << digit)) != 0) {
                return false;
            }
            mask |= 1 << digit;
        }
        return mask == 0x3FE;
    }
}
